tensornetwork/quantum/quantum.py

```diff
@@ -0,0 +1,374 @@
+
+    return quantum_constructor(out_edges, in_edges, ref_nodes, ignore_edges)
+
+  def __mul__(self, other: "QuOperator"):
+    if self.is_scalar or other.is_scalar:
+      return self @ other
+    raise ValueError("Elementwise multiplication is only supported if at "
+                     "least one of the arguments is a scalar.")
+
+  def tensor_prod(self, other: "QuOperator"):
+    """Tensor product with another operator.
+    """
+    nodes_dict1, edges_dict1 = copy(self.nodes, False)
+    nodes_dict2, edges_dict2 = copy(other.nodes, False)
+
+    in_edges = ([edges_dict1[e] for e in self.in_edges] +
+                [edges_dict2[e] for e in other.in_edges])
+    out_edges = ([edges_dict1[e] for e in self.out_edges] +
+                 [edges_dict2[e] for e in other.out_edges])
+    ref_nodes = ([n for _, n in nodes_dict1.items()] +
+                 [n for _, n in nodes_dict2.items()])
+    ignore_edges = ([edges_dict1[e] for e in self.ignore_edges] +
+                    [edges_dict2[e] for e in other.ignore_edges])
+
+    return quantum_constructor(out_edges, in_edges, ref_nodes, ignore_edges)
+
+  def contract(self, contractor: Callable = greedy,
+               final_edge_order: Optional[Sequence[Edge]] = None):
+    """Contract the tensor network in place.
+
+    This modifies the tensor network representation of the operator (or vector,
+    or scalar), reducing it to a single tensor, without changing the value.
+
+    Args:
+      contractor: A function that performs the contraction. Defaults to
+        `greedy`, which uses the greedy algorithm from `opt_einsum` to
+        determine a contraction order.
+      final_edge_order: Manually specify the axis ordering of the final tensor.
+    Returns:
+      The present object.
+    """
+    if final_edge_order:
+      self.ref_nodes = set(
+          [contractor(self.nodes, output_edge_order=final_edge_order)])
+    else:
+      self.ref_nodes = set([contractor(self.nodes, ignore_edge_order=True)])
+    return self
+
+  def eval(self, contractor: Callable = greedy,
+           final_edge_order: Optional[Sequence[Edge]] = None):
+    """Contracts the tensor network in place and returns the final tensor.
+
+    Note that this modifies the tensor network representing the operator.
+
+    The default ordering for the axes of the final tensor is:
+      `out_edges + in_edges`
+    If there are any "ignored" edges, their axes come first:
+      `ignored_edges + out_edges + in_edges`
+
+    Args:
+      contractor: A function that performs the contraction. Defaults to
+        `greedy`, which uses the greedy algorithm from `opt_einsum` to
+        determine a contraction order.
+      final_edge_order: Manually specify the axis ordering of the final tensor.
+    Returns:
+      The final tensor representing the operator.
+    """
+    if not final_edge_order:
+      # TODO: Perhaps support using no assigned ordering at all here?
```

Comment:
I think we should always force a user to give the final edge order or have them explicitly turn that off like we do with `contractors`.

Reply:
In this case there is an order implied by `in_edges` and `out_edges` that I think we can safely make the default (this is the default ordering I have picked). Given row-major memory layout, it is the ordering corresponding to the Kronecker product.